### PR TITLE
Fixed #32744 -- Normalized to pathlib.Path in autoreloader check for template changes.

### PR DESCRIPTION
**Pull request by hramezani**
https://code.djangoproject.com/ticket/32744

### CODE REVIEW

**Comment by hramezani** [comment]
I added the second commit to change tests and then tests passed. with the original tests(first commit) we had this error on `windows`:
```
Traceback (most recent call last):
  File "C:\Jenkins\workspace\pull-requests-windows\database\sqlite3\label\windows\python\Python39\django\test\utils.py", line 430, in inner
    return func(*args, **kwargs)
  File "C:\Jenkins\workspace\pull-requests-windows\database\sqlite3\label\windows\python\Python39\tests\template_tests\test_autoreloader.py", line 77, in test_get_template_directories_dirs
    self.assertSetEqual(
AssertionError: Items in the first set but not the second:
WindowsPath('template_tests/template_dir')
WindowsPath('template_tests/template_dir1')
Items in the second set but not the first:
WindowsPath('C:/Jenkins/workspace/pull-requests-windows/database/sqlite3/label/windows/python/Python39/tests/template_tests/template_dir1')
WindowsPath('C:/Jenkins/workspace/pull-requests-windows/database/sqlite3/label/windows/python/Python39/tests/template_tests/template_dir')
```

**Review by carltongibson — APPROVED** [review]
Hi @hramezani — Thanks for this. I adjusted to include each of the cases @felixxm [mentioned on the ticket](https://code.djangoproject.com/ticket/32744#comment:6) It's good because without the `resolve()` call even `PosixPath('template_tests/relative_path')` isn't as expected. What do you think? 

Hey @rpkilby 👋 — this should work for you. Fancy giving it a run to confirm? 🏅

**Comment by carltongibson** [comment]
Grrr. @hramezani I need to have a play with this failure — what do you think? 

```
AssertionError: Items in the first set but not the second:
WindowsPath('C:/Jenkins/workspace/pull-requests-windows/database/sqlite3/label/windows/python/Python39/template_tests/relative_path')
WindowsPath('C:/Jenkins/workspace/pull-requests-windows/database/sqlite3/label/windows/python/Python39/template_tests/relative_str')
Items in the second set but not the first:
WindowsPath('C:/Jenkins/workspace/pull-requests-windows/database/sqlite3/label/windows/python/Python39/tests/template_tests/relative_path')
WindowsPath('C:/Jenkins/workspace/pull-requests-windows/database/sqlite3/label/windows/python/Python39/tests/template_tests/relative_str')
```

The extra `tests` is, what 🤔, the root folder for the relative cases not being the same as the linux builds.

**Comment by carltongibson** [comment]
@hramezani I'll take a look at it in the morning. Thanks.

**Comment by carltongibson** [comment]
The Windows failure depends on `pwd`. From the repo root the test fails. From the `tests` folder it passes.

File: django/template/autoreload.py [hunk 1]
```diff
@@ -16,13 +17,13 @@ def get_template_directories():
         if not isinstance(backend, DjangoTemplates):
             continue
 
-        items.update(to_path(dir).absolute() for dir in backend.engine.dirs)
+        items.update(Path.cwd() / to_path(dir) for dir in backend.engine.dirs)
```
Comment (felixxm) [inline django/template/autoreload.py:20]: Why not `resolve()`? :thinking:

Comment (carltongibson) [reply django/template/autoreload.py:20]: It doesn't work on Windows. :| Non-existent files simple **don't** get resolved. 

```
>>> from pathlib import Path
>>> file = Path('new.txt')
>>> file.exists()
False
>>> file.resolve()
WindowsPath('new.txt')
>>> file.absolute()
WindowsPath('d:/home/new.txt')
>>> file.touch()
>>> file.resolve()
WindowsPath('D:/home/new.txt')
>>> file.unlink()
>>> file.resolve()
WindowsPath('new.txt')
```

This is presumably why @hramezani went for `absolute()` (right?) but that's undocumented and shouldn't be used. 

> Path.absolute() is undocumented and thus unsupported, so use at your own peril. — Brett Cannon. 

* https://discuss.python.org/t/pathlib-absolute-vs-resolve/2573
* https://bugs.python.org/issue38671
* https://bugs.python.org/issue39090
* **Slightly Misleading in that it suggests `absolute()`**: https://stackoverflow.com/questions/42513056/how-to-get-absolute-path-of-a-pathlib-path-object

File: tests/template_tests/test_autoreloader.py [hunk 2]
```diff
@@ -64,6 +64,26 @@ def test_reset_all_loaders(self, mock_reset):
...
+            'BACKEND': 'django.template.backends.django.DjangoTemplates',
+        }]
+    )
+    def test_template_dirs_normalized_to_paths(self):
+        self.assertSetEqual(
+            autoreload.get_template_directories(),
+            {
+                ROOT / 'absolute_str',
+                Path.cwd() / 'template_tests/relative_str',
+                Path.cwd() / 'template_tests/relative_path',
```
Comment (carltongibson) [inline tests/template_tests/test_autoreloader.py:83]: So, I think we need a comment here, about why the difference.

File: django/template/autoreload.py (hunk 1 above)
Comment (felixxm) [reply django/template/autoreload.py:20]: But should we worry about non-existent directories? we can use `template_tests/templates/first` and `template_tests/templates/second` in tests.

**Review by carltongibson — COMMENTED** [review]
So assuming CI passes 🤞 I think this is how we need to do it. 

`Path.absolute()` is undocumented and `Path.resolve()` doesn't work for non-existent files on Windows. :| (`Path.cwd() / relative_path` is the recommended way to do it, apparently.)

**Comment by carltongibson** [comment]
> So assuming CI passes 🤞

And of course it doesn't. 

So locally, this at 997a442 passes running from the root dir and `tests`. Why not for Jenkins? 🤔

**Comment by carltongibson** [comment]
It's slightly frustrating now, but the import of the test is just that the `DIRS` were all converted to `Path` instances so we could just assert that. Leaving that there's some weirdness in the actual paths generated from relative values. (The docs don't say that's not allowed, but it's surely not a good idea...)

File: django/template/autoreload.py (hunk 1 above)
Comment (carltongibson) [reply django/template/autoreload.py:20]: They'll still be dependent on `cwd` I think... 🤔

**Comment by carltongibson** [comment]
Aha! 👯 When in _This should be passing mode_, leave it over the weekend, rebase, and see. 9 times in 10 just what's needed https://djangoci.com/job/pull-requests-windows/database=sqlite3,label=windows,python=Python39/11578/

(Now just hoping the rest still pass...)

**Review by carltongibson — APPROVED** [review]

**Comment by urfinorg** [comment]
Hello! Maybe i miss something, but if 'APP_DIRS': True, autoreload will not work with ANY files(views.py, ...) inside apps. As i can see, it absulutly broke autoreload logic.  I'm facing with it on django==3.2.4 and next. On 3.2.3 everything works as properly.

**Pull request merged**